include/seqan3/range/view/single_pass_input.hpp

```diff
@@ -0,0 +1,402 @@
+    using const_iterator    = iterator;
+    //\}
+
+    /*!\name Constructor, destructor, and assignment.
+     * \{
+     * \brief All standard functions are explicitly defaulted.
+     */
+    constexpr single_pass_input_view() = default;
+    constexpr single_pass_input_view(single_pass_input_view const &) = default;
+    constexpr single_pass_input_view(single_pass_input_view &&) = default;
+    constexpr single_pass_input_view & operator=(single_pass_input_view const &) = default;
+    constexpr single_pass_input_view & operator=(single_pass_input_view &&) = default;
+    ~single_pass_input_view() = default;
+
+    //!\brief Construction from the underlying view.
+    single_pass_input_view(urng_t && urng) :
+        state_ptr{new state{std::forward<urng_t>(urng), urng.begin()}}
+    {}
+    //!\}
+
+    /*!\name Iterators
+     * \{
+     */
+    /*!\brief Returns an iterator to the current begin of the underlying range.
+     *
+     * \details
+     *
+     * Subsequent calls to begin will result in different positions if the iterator was incremented
+     * between the calls.
+     */
+    iterator begin()
+    {
+        return iterator{*this};
+    }
+
+    //!\brief Const version of begin is deleted, since the underlying state must be mutable.
+    const_iterator begin() const = delete;
+
+    //!\copydoc single_pass_input_view::begin() const
+    const_iterator cbegin() const = delete;
+
+    //!\brief Returns a sentinel.
+    sentinel end()
+    {
+        // calling to urng->end must be in O(1).
+        return sentinel{state_ptr->urng.end()};
+    }
+
+    //!\brief Const version of end is deleted, since the underlying state must be mutable.
+    sentinel end() const = delete;
+
+    //!\copydoc single_pass_input_view::end() const
+    sentinel cend() const = delete;
+    //!\}
+};
+
+/*!\name Deduction guide.
+ * \relates seqan3::detail::single_pass_input_view
+ * \{
+ */
+//!\brief Deduces the single_pass_input_view from the underlying range.
+template <input_range_concept urng_t>
+single_pass_input_view(urng_t &&) -> single_pass_input_view<urng_t>;
+//!\}
+} // seqan3::detail
+
+//-----------------------------------------------------------------------------
+// Iterator for single pass input view.
+//-----------------------------------------------------------------------------
+
+namespace seqan3::detail
+{
+/*!\brief An input_iterator over the associated range.
+ * \implements input_iterator_concept
+ * \ingroup view
+ * \tparam view_type The type of the associated type.
+ *
+ * This iterator reduces every iterator type of the associated view to an single pass input iterator.
+ */
+template <typename view_type>
+class single_pass_input_iterator<single_pass_input_view<view_type>>
+{
+    //!\brief The pointer to the associated view.
+    using base_iterator_t = typename single_pass_input_view<view_type>::urng_iterator_type;
+    //!\brief The pointer to the associated view.
+    single_pass_input_view<view_type> * view_ptr{};
```

Comment:
same here, I would propose removing `_ptr`. This might have made sense when using owning raw pointers to highlight "REMEMBER TO DELETE ME", but for non-owning and smart pointers, I don't think this makes sense.

Reply:
It is so much more expressive than just calling it view or something else. It is a pointer to the view, which clearly expresses the way how to interact with it. 
BTW, owning-raw pointers not wrapped in a RAII class would not be allowed anyway.

Reply:
> It is so much more expressive than just calling it view or something else. It is a pointer to the view, which clearly expresses the way how to interact with it.

We never encode the type inside the typename, although there are many conventions to do so:
```cpp
int m_foobar; // for member variables
static int s_foobar; // for static member variables

int m_i_foobar; // contains i for int
```
All of these are more "expressive" than what we are doing, all of them arguably express more clearly of how to interact with the variable; *but we are not doing this anywhere else*.

In the `view` you have `urng`. Is this a variable, is it a pointer, is it a reference? The variable name doesn't tell, and it is not even fixed (could be value or reference). I am not saying it is bad in general, it's just very different from our other style.

Reply:
Well, `rng` is perfectly the same thing. It tells me that the referred object is a range (or mistakenly a random number generator). It would be even better to call it `underlying_rng`. So what is the difference to `..._ptr`?